### DISCUSSION TITLE
How to use queries?

### DISCUSSION THREAD

**Discussion by MarErm27** [slick/slick#2407] [Questions] [answered]
For example, I run a site with science posts and I have to approve every post before publishing.

Step 1
I have a sequence of post ids that I have approved, and I want to determine if they are fresh. I call a query that checks if a field posts.createdAt is today, then they are fresh otherwise they are old.

Step 2
For fresh posts, I want to retrieve post's titles, descriptions and user's names.
For old posts, I want to retrieve only titles, user's name and createdAt without description.

Step 3
Then I store all results in the database in a table called today's posts that I use as a cache. This table will be called whenever a user opens a main page of my site.

I can combine step 1 and step 2 to one big query. Instead, I want to keep the logic of each step separated to make it maintainable and reusable.

Here is pseudocode of something that I expect to see:
```
def collectTodayPost(ids: Seq[Int]) = {
    for {
        isFresh:Seq[Int, Boolean] <- DAO.isFresh(ids)
        todayPosts <- DAO.getTodayPosts(isFresh)
        _ <- DAO.writeTodayPosts(todayPosts)
    }
}
```

```
package DAO

def isFresh(ids:Seq[Int]) = {
    val query = for {
        isFresh <- Posts.collect(if createdAt == today true else false)
    } yield Seq(Int, Boolean)
    db run query.result
}

def getTodayPosts(isFresh: Seq[Int,Boolean]) = {
    val query = for {
        posts <- isFresh match {
            case True => getFreshPosts(Seq[Ids])
            case False => getOldPosts(Seq[Ids])
        } yield posts
    }
    db run query posts.result
}
```

This is a pseudocode because I haven't yet found a way to send the result of isFresh function to getTodayPosts functions and how to match isFresh in getTodayPosts.

The problem of this decision is that isFresh have the "db run query.result" calls which is a call to the database and getTodayPosts also have it. getFreshPosts and getOldPosts also suppose to have some db calls and it created many calls. I'm looking for a way to reduce the number of calls to the database to 1.

**Comment by MarErm27** [comment]
I've added a repository where I'm trying to accomplish the goals
https://github.com/MarErm27/scienceExample/blob/main/src/main/scala/com/howtographql/scala/sangria/DAO.scala
I'm trying to make these three functions work, may you recommend me how to do it?
collectTodayPost
isFreshPost
getTodayPosts

**Comment by nafg** [comment, marked as answer]
I forked it and got it to compile: https://github.com/nafg/scienceExample

To compose queries, I'm not sure the right way but here is one way. Not sure if it will have the semantics you want: https://github.com/nafg/scienceExample/tree/compose-query

**Comment by MarErm27** [comment]
@nafg Thank you!
The result of getTodayPosts is a sequence of tuples and if I want to get access to its fields I have to remember their order, so it looks like this _._1, _._2 ...
Is there a way to call this fields by names?
I'm trying to convert this sequence of tuples to hashmap of hashmaps here:
https://github.com/MarErm27/scienceExample/blob/a42f919094dabd799f8a8ce231dd4cb7853a8699/src/main/scala/com/howtographql/scala/sangria/DAO.scala#L136

but it says ""No implicits found for parameter shape: Shape[_ <: FlatShapeLevel, HashMap[Rep[Int], HashMap[String, Rep[_ >: Int with String with DateTime]]], T_, G_]""

How to call fields of the result by names?

**Reply by nafg** [reply]
Before `.result` you're writing code that is not actually logic that runs, but is "introspected" to build a query. The error means that Slick doesn't know how to write a query expression that builds a HashMap. Neither do I.

Perhaps you want to transform the data after it comes back from the database to the application. In that case, you want to transform the DBIO, not the Query. Use a (nested) `.map` _after_ `.result`.
